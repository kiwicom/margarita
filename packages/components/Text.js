// @flow

import * as React from 'react';
import { Text as RNText } from 'react-native';
import { StyleSheet } from '@kiwicom/universal-components';

const Text = ({ children }: { children: React.Node }) => {
  return <RNText style={styles.text}>{children}</RNText>;
};

const styles = StyleSheet.create({
  text: {
    alignItems: 'center',
    fontSize: 24,
  },
});

export default Text;
