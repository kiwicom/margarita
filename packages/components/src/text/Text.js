// @flow

import * as React from 'react';
import { Platform } from 'react-native';
import {
  Text as UniversalComponentsText,
  StyleSheet,
} from '@kiwicom/universal-components';

/**
 * This wrapper is necessary because loading fonts with Expo does not allow using fontWeight and fontStyle
 * cf. https://docs.expo.io/versions/latest/guides/using-custom-fonts/
 */

export default function Text(
  props: $PropertyType<typeof UniversalComponentsText, 'props'>,
) {
  if (Platform.OS !== 'ios' && Platform.OS !== 'android') {
    return <UniversalComponentsText {...props} />;
  }
  const { italic, weight } = props;

  let fontFamily = styles.normal;
  if (italic && weight === 'bold') {
    fontFamily = styles.boldItalic;
  } else if (italic) {
    fontFamily = styles.italic;
  } else if (weight === 'bold') {
    fontFamily = styles.bold;
  }

  return (
    <UniversalComponentsText {...props} style={[props.style, fontFamily]} />
  );
}

const styles = StyleSheet.create({
  normal: {
    fontFamily: 'Roboto',
  },
  bold: {
    fontFamily: 'RobotoBold',
  },
  italic: {
    fontFamily: 'RobotoItalic',
  },
  boldItalic: {
    fontFamily: 'RobotoBoldItalic',
  },
});
