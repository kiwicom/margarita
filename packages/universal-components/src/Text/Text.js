// @flow

import * as React from 'react';
import { Text as RNText, Platform } from 'react-native';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';

import { StyleSheet } from '../PlatformStyleSheet';
import { createStylesGenerator } from '../utils';
import {
  align as alignTypes,
  fontSize,
  fontWeight as fontWeightTypes,
  textColor,
} from './styles';
import type { TextType } from './TextTypes';

const colorGen = createStylesGenerator('color', textColor);
const fontWeightGen = createStylesGenerator('fontWeight', fontWeightTypes);
const fontSizeGen = createStylesGenerator('fontSize', fontSize);
const alignGen = createStylesGenerator('textAlign', alignTypes);

const Text = ({
  align,
  children,
  dataTest,
  italic,
  numberOfLines,
  size,
  style,
  type,
  uppercase,
  weight,
  expo,
}: TextType) => {
  const textStyle = [styles.text];
  if (italic) {
    textStyle.push(styles.italic);
  }
  if (uppercase) {
    textStyle.push(styles.uppercase);
  }
  if (weight) {
    textStyle.push(styles[weight]);
  }
  if (size) {
    textStyle.push(styles[size]);
  }
  if (type) {
    textStyle.push(styles[type]);
  }
  if (align) {
    textStyle.push(styles[align]);
  }
  if (style) {
    textStyle.push(style);
  }

  if (expo && (Platform.OS === 'ios' || Platform.OS === 'android')) {
    let fontFamily = styles.normalFontFamily;
    if (italic && weight === 'bold') {
      fontFamily = styles.boldItalicFontFamily;
    } else if (italic) {
      fontFamily = styles.italicFontFamily;
    } else if (weight === 'bold') {
      fontFamily = styles.boldFontFamily;
    }
    textStyle.push(fontFamily);
  }

  return (
    <RNText
      data-test={dataTest}
      style={textStyle}
      numberOfLines={numberOfLines}
    >
      {children}
    </RNText>
  );
};

Text.defaultProps = {
  expo: false,
};

const styles = StyleSheet.create({
  text: {
    margin: 0,
    fontFamily: Platform.OS === 'web' ? defaultTokens.fontFamily : 'Roboto',
  },
  italic: {
    fontStyle: 'italic',
  },
  uppercase: {
    textTransform: 'uppercase',
  },
  ...fontWeightGen(),
  ...alignGen(),
  ...colorGen(),
  ...fontSizeGen(),
  normalFontFamily: {
    fontFamily: 'Roboto',
  },
  boldFontFamily: {
    fontFamily: 'RobotoBold',
  },
  italicFontFamily: {
    fontFamily: 'RobotoItalic',
  },
  boldItalicFontFamily: {
    fontFamily: 'RobotoBoldItalic',
  },
});

export default Text;
