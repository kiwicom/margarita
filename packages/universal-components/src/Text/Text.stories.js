// @flow

import React from 'react';
import { View } from 'react-native';
import { storiesOf } from '@storybook/react-native';
import { boolean, select, withKnobs } from '@storybook/addon-knobs';

import { Text } from '.';

const style = {
  alignSelf: 'stretch',
  flexGrow: 1,
  justifyContent: 'center',
  alignItems: 'center',
  padding: 25,
  backgroundColor: '#f5fcff',
  width: '100%',
};
const customText =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse iaculis, nulla at luctus ultrices, dolor.';

storiesOf('Text', module)
  .addDecorator(withKnobs)
  .addDecorator(getStory => <View style={style}>{getStory()}</View>)
  .add('Cascading styles', () => (
    <Text style={{ color: 'blue', fontSize: 22 }}>
      Test
      <Text style={{ color: 'red' }}>Nested test</Text>
    </Text>
  ))
  .add('Primary text', () => <Text type="primary">{customText}</Text>)
  .add('Secondary text', () => <Text type="secondary">{customText}</Text>)
  .add('Attention text', () => <Text type="attention">{customText}</Text>)
  .add('Status text', () => (
    <React.Fragment>
      <Text type="info">{customText}</Text>
      <Text type="success">{customText}</Text>
      <Text type="warning">{customText}</Text>
      <Text type="critical">{customText}</Text>
    </React.Fragment>
  ))
  .add('Playground', () => {
    const type = select(
      'Type',
      [
        'primary',
        'secondary',
        'attention',
        'info',
        'success',
        'warning',
        'critical',
        'white',
      ],
      'primary',
    );
    const size = select('Size', ['small', 'normal', 'large'], 'small');
    const weight = select('Weight', ['normal', 'bold'], 'normal');
    const align = select(
      'Align',
      ['left', 'right', 'center', 'justify'],
      'right',
    );
    const uppercase = boolean('Uppercase', false);
    const italic = boolean('Italic', false);

    return (
      <Text
        type={type}
        size={size}
        weight={weight}
        align={align}
        uppercase={uppercase}
        italic={italic}
      >
        {customText}
      </Text>
    );
  });
