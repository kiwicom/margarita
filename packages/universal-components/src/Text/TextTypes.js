// @flow

import * as React from 'react';

import { type StylePropType } from '../PlatformStyleSheet/StyleTypes';
import { type Theme } from '../ThemeProvider';

export type TextType = {|
  align?: 'left' | 'right' | 'center' | 'justify',
  +children: ?React.Node,
  +dataTest?: string,
  +italic?: boolean,
  numberOfLines?: ?number,
  ellipsizeMode?: 'head' | 'middle' | 'tail' | 'clip',
  size?: 'normal' | 'large' | 'small',
  +style?: StylePropType,
  +uppercase?: boolean,
  +type?:
    | 'attention'
    | 'critical'
    | 'info'
    | 'primary'
    | 'secondary'
    | 'success'
    | 'warning'
    | 'white',
  +weight?: 'normal' | 'bold',
  +expo?: boolean,
  +theme: Theme,
|};
