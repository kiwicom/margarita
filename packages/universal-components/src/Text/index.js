// @flow

export { default as Text } from './Text';
