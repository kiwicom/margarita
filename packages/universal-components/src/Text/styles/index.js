// @flow

import { defaultTokens } from '@kiwicom/orbit-design-tokens';

export const textColor = {
  attention: defaultTokens.colorTextAttention,
  critical: defaultTokens.colorTextCritical,
  info: defaultTokens.colorTextInfo,
  primary: defaultTokens.colorTextPrimary,
  secondary: defaultTokens.colorTextSecondary,
  success: defaultTokens.colorTextSuccess,
  warning: defaultTokens.colorTextWarning,
  white: defaultTokens.colorTextWhite,
};

export const fontSize = {
  large: parseFloat(defaultTokens.fontSizeTextLarge),
  normal: parseFloat(defaultTokens.fontSizeTextNormal),
  small: parseFloat(defaultTokens.fontSizeTextSmall),
};

export const fontWeight = {
  normal: defaultTokens.fontWeightNormal,
  bold: defaultTokens.fontWeightBold,
};

export const align = {
  center: 'center',
  justify: 'justify',
  left: 'left',
  right: 'right',
};
