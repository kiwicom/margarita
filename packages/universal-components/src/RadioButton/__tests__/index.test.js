// @flow

import * as React from 'react';
import { render, fireEvent } from 'react-native-testing-library';
import snapshotDiff from 'snapshot-diff';

import { RadioButton } from '..';

import { Text } from '../../Text';

describe('RadioButton', () => {
  const type = 'bullet';
  const bulletPosition = 'left';
  const checked = true;
  const disabled = false;
  const label = 'Label';
  const children = <Text>child-label</Text>;
  const onPress = jest.fn();
  const style = { padding: 10 };

  const { getByType, getByText, getAllByProps } = render(
    <RadioButton
      type={type}
      bulletPosition={bulletPosition}
      checked={checked}
      disabled={disabled}
      onPress={onPress}
      style={style}
      label={label}
    >
      {children}
    </RadioButton>,
  );

  it('should contain child', () => {
    expect(getByText('child-label')).toBeDefined();
  });

  it('should render children rather than label', () => {
    expect(() => {
      getByText(label);
    }).toThrow();
  });

  it('should have passed props', () => {
    expect(
      getAllByProps({
        type,
        bulletPosition,
        checked,
        disabled,
        children,
        onPress,
        style,
        label,
      }),
    ).toBeDefined();
  });

  it('should execute onPress method', () => {
    fireEvent(getByType(RadioButton), 'press');
  });

  it('should match snapshot diff', () => {
    const base = <RadioButton onPress={onPress}>{children}</RadioButton>;
    const extend = (
      <RadioButton
        type={type}
        bulletPosition={bulletPosition}
        checked={checked}
        disabled={disabled}
        onPress={onPress}
        style={style}
      >
        {children}
      </RadioButton>
    );

    expect(snapshotDiff(base, extend)).toMatchSnapshot();
  });

  it('should render label', () => {
    const wrapper = render(<RadioButton onPress={onPress} label="label" />);
    expect(wrapper.getByTestId('label-wrapper')).toBeDefined();
  });
});
