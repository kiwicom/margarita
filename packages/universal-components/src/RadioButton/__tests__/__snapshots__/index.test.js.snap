// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton should match snapshot diff 1`] = `
"Snapshot Diff:
- <RadioButton onPress={[Function mockConstructor]}><Text expo={false}>child-label</Text></RadioButton>
+ <RadioButton bulletPosition=\\"left\\" checked={true} disabled={false} onPress={[Function mockConstructor]} style={{\\"padding\\": 10}} type=\\"bullet\\"><Text expo={false}>child-label</Text></RadioButton>

@@ -9,12 +9,14 @@
    style={
      Array [
        Object {
          \\"flexDirection\\": \\"row\\",
        },
-       undefined,
-       undefined,
+       false,
+       Object {
+         \\"padding\\": 10,
+       },
      ]
    }
  >
    <View
      style={
@@ -26,15 +28,28 @@
            \\"borderWidth\\": 2,
            \\"height\\": 24,
            \\"justifyContent\\": \\"center\\",
            \\"width\\": 24,
          },
-         undefined,
-         undefined,
+         Object {
+           \\"borderColor\\": \\"#00a991\\",
+         },
+         false,
        ]
      }
+   >
+     <View
+       style={
+         Object {
+           \\"backgroundColor\\": \\"#00a991\\",
+           \\"borderRadius\\": 7,
+           \\"height\\": 14,
+           \\"width\\": 14,
+         }
+       }
      />
+   </View>
    <Text
      style={
        Array [
          Object {
            \\"fontFamily\\": \\"Roboto\\","
`;
